File: dbg/utils/lpc.py
import math
import numpy as np
from typing import List, Optional
# from alphas_plot import analyse_alphas

class LPC:
    def __init__(self, num_channels: int):
        self.total_num_channels = num_channels
        self.FRAMELEN = 441
        self.prev_frame_len = self.FRAMELEN
        self.HOPSIZE = self.FRAMELEN // 2
        self.BUFLEN = 16384
        self.SAMPLERATE = 44100
        self.MAX_EXLEN = self.SAMPLERATE // 6
        self.EXLEN = self.MAX_EXLEN
        self.ORDER = 32
        self.ex_type = 0
        self.order_changed = False
        self.ex_type_changed = False
        self.ex_start_changed = False
        self.ex_start = 0.0
        self.start = False
        self.noise = None
        self.G = 1
        
        # Initialize pointers
        self.in_wt_ptrs = [0] * num_channels
        self.in_rd_ptrs = [0] * num_channels
        self.smp_cnts = [0] * num_channels
        self.out_wt_ptrs = [self.HOPSIZE] * num_channels
        self.out_rd_ptrs = [0] * num_channels
        self.ex_ptrs = [0] * num_channels
        self.hist_ptrs = [0] * num_channels
        self.ex_cnt_ptrs = [0] * num_channels
        
        # Initialize buffers
        self.phi = np.zeros(self.ORDER + 1)
        self.alphas = np.zeros(self.ORDER + 1)
        self.reflections = np.zeros(self.ORDER)
        
        ordered_buf_size = int(self.SAMPLERATE * 0.01)
        self.ordered_in_buf = np.zeros(ordered_buf_size)
        self.ordered_sc_buf = np.zeros(ordered_buf_size)
        self.window = np.zeros(ordered_buf_size)
        
        # Initialize channel-specific buffers
        self.in_buf = [np.zeros(self.BUFLEN) for _ in range(num_channels)]
        self.sc_buf = [np.zeros(self.BUFLEN) for _ in range(num_channels)]
        self.out_buf = [np.zeros(self.BUFLEN) for _ in range(num_channels)]
        self.out_hist = [np.zeros(self.ORDER) for _ in range(num_channels)]
        
        # Initialize window
        for i in range(self.FRAMELEN):
            self.window[i] = 0.5 * (1.0 - math.cos(2.0 * math.pi * i / (self.FRAMELEN - 1)))
        
        self.reset_a()
    
    def reset_a(self):
        self.alphas[0] = 1.0
        for i in range(1, self.ORDER + 1):
            self.alphas[i] = 0.0
    
    def levinson_durbin(self):
        self.reset_a()
        E = self.phi[0]
        
        for k in range(self.ORDER):
            lbda = 0.0
            for j in range(k + 1):
                lbda += self.alphas[j] * self.phi[k + 1 - j]
            
            lbda = -lbda / E
            self.reflections[k] = lbda
            
            half = (k + 1) // 2
            for n in range(half + 1):
                tmp = self.alphas[k + 1 - n] + lbda * self.alphas[n]
                self.alphas[n] += lbda * self.alphas[k + 1 - n]
                self.alphas[k + 1 - n] = tmp
            
            E *= (1.0 - lbda * lbda)
    
    def autocorrelate(self, x: np.ndarray, frame_size: int, lag: int) -> float:
        res = 0.0
        for n in range(frame_size - lag):
            res += x[n] * x[n + lag]
        return res
    
    def prepare_to_play(self):
        self.in_wt_ptrs = [0] * self.total_num_channels
        self.in_rd_ptrs = [0] * self.total_num_channels
        self.smp_cnts = [0] * self.total_num_channels
        self.out_wt_ptrs = [self.HOPSIZE] * self.total_num_channels
        self.out_rd_ptrs = [0] * self.total_num_channels
        self.ex_ptrs = [0] * self.total_num_channels
        self.hist_ptrs = [0] * self.total_num_channels
        self.ex_cnt_ptrs = [0] * self.total_num_channels
        
        self.in_buf = [np.zeros(self.BUFLEN) for _ in range(self.total_num_channels)]
        self.out_buf = [np.zeros(self.BUFLEN) for _ in range(self.total_num_channels)]
        self.out_hist = [np.zeros(self.ORDER) for _ in range(self.total_num_channels)]
        self.sc_buf = [np.zeros(self.BUFLEN) for _ in range(self.total_num_channels)]
    
    def apply_lpc(self, input_signal: np.ndarray, output_signal: np.ndarray, num_samples: int, 
                 lpc_mix: float, ex_percentage: float, ch: int, ex_start_pos: float, 
                 sidechain: Optional[np.ndarray] = None, previous_gain: float = 1.0, 
                 current_gain: float = 1.0):
        
        in_wt_ptr = self.in_wt_ptrs[ch]
        in_rd_ptr = self.in_rd_ptrs[ch]
        smp_cnt = self.smp_cnts[ch]
        out_rd_ptr = self.out_rd_ptrs[ch]
        out_wt_ptr = (out_rd_ptr + self.HOPSIZE) % self.BUFLEN
        
        if self.HOPSIZE < self.prev_frame_len // 2:
            for i in range(self.prev_frame_len // 2 - self.HOPSIZE):
                idx = out_wt_ptr + i
                if idx >= self.BUFLEN:
                    idx -= self.BUFLEN
                self.out_buf[ch][idx] = 0
        
        ex_start = int(ex_start_pos * self.EXLEN)
        
        if self.ex_type_changed:
            self.ex_ptrs[ch] = ex_start
            self.ex_cnt_ptrs[ch] = 0
            self.hist_ptrs[ch] = 0
        
        if self.order_changed:
            self.out_hist[ch].fill(0)
            self.hist_ptrs[ch] = 0
        
        ex_ptr = self.ex_ptrs[ch]
        ex_cnt_ptr = self.ex_cnt_ptrs[ch]
        hist_ptr = self.hist_ptrs[ch]
        
        slope = (current_gain - previous_gain) / num_samples
        valid_hop_size = min(self.HOPSIZE, self.prev_frame_len // 2)
        
        for s in range(num_samples):
            self.in_buf[ch][in_wt_ptr] = input_signal[s]
            
            if sidechain is not None:
                self.sc_buf[ch][in_wt_ptr] = sidechain[s]
            
            in_wt_ptr += 1
            if in_wt_ptr >= self.BUFLEN:
                in_wt_ptr = 0
            
            out = self.out_buf[ch][out_rd_ptr]
            in_val = self.in_buf[ch][(in_rd_ptr + self.BUFLEN - self.FRAMELEN) % self.BUFLEN]
            
            in_rd_ptr += 1
            if in_rd_ptr >= self.BUFLEN:
                in_rd_ptr = 0
            
            gain_factor = previous_gain + slope * float(s)
            output_signal[s] = lpc_mix * gain_factor * out + (1 - lpc_mix) * in_val
            self.out_buf[ch][out_rd_ptr] = 0
            
            out_rd_ptr += 1
            if out_rd_ptr >= self.BUFLEN:
                out_rd_ptr = 0
            
            smp_cnt += 1
            
            if smp_cnt >= valid_hop_size:
                smp_cnt = 0
                
                if sidechain is not None:
                    for i in range(self.FRAMELEN):
                        in_buf_idx = (in_wt_ptr + i - self.FRAMELEN + self.BUFLEN) % self.BUFLEN
                        self.ordered_in_buf[i] = self.window[i] * self.in_buf[ch][in_buf_idx]
                        self.ordered_sc_buf[i] = self.sc_buf[ch][in_buf_idx]
                else:
                    for i in range(self.FRAMELEN):
                        in_buf_idx = (in_wt_ptr + i - self.FRAMELEN + self.BUFLEN) % self.BUFLEN
                        self.ordered_in_buf[i] = self.window[i] * self.in_buf[ch][in_buf_idx]
                
                for lag in range(self.ORDER + 1):
                    self.phi[lag] = self.autocorrelate(self.ordered_in_buf, self.FRAMELEN, lag)
                
                if self.phi[0] != 0:
                    self.levinson_durbin()
                    
                    self.G = self.phi[0]
                    for k in range(self.ORDER):
                        self.G -= self.alphas[k + 1] * self.phi[k + 1]
                    
                    # self.G = math.sqrt(G / self.FRAMELEN / math.sqrt(float(self.ORDER)))
                    self.G = math.sqrt(self.G)
                    if sidechain is None:
                        for n in range(self.FRAMELEN):
                            ex = self.noise[ex_ptr]
                            ex_cnt_ptr += 1
                            ex_ptr += 1
                            
                            if ex_cnt_ptr >= int(ex_percentage * self.EXLEN):
                                ex_cnt_ptr = 0
                                ex_ptr = ex_start
                            
                            if ex_ptr >= self.EXLEN:
                                ex_ptr = 0
                            
                            out_n = G * ex
                            for k in range(self.ORDER):
                                idx = (hist_ptr + k) % self.ORDER
                                out_n -= self.alphas[k + 1] * self.out_hist[ch][idx]
                            
                            hist_ptr -= 1
                            if hist_ptr < 0:
                                hist_ptr += self.ORDER
                            
                            self.out_hist[ch][hist_ptr] = out_n
                            wt_idx = (out_wt_ptr + n) % self.BUFLEN
                            self.out_buf[ch][wt_idx] += out_n
                    else:
                        for n in range(self.FRAMELEN):
                            ex = self.ordered_sc_buf[n]
                            ex_cnt_ptr += 1
                            ex_ptr += 1
                            
                            if ex_cnt_ptr >= int(ex_percentage * self.EXLEN):
                                ex_cnt_ptr = 0
                                ex_ptr = ex_start
                            
                            if ex_ptr >= self.EXLEN:
                                ex_ptr = 0
                            
                            out_n = self.G * ex
                            for k in range(self.ORDER):
                                idx = (hist_ptr + k) % self.ORDER
                                out_n -= self.alphas[k + 1] * self.out_hist[ch][idx]
                            
                            hist_ptr -= 1
                            if hist_ptr < 0:
                                hist_ptr += self.ORDER
                            
                            self.out_hist[ch][hist_ptr] = out_n
                            wt_idx = (out_wt_ptr + n) % self.BUFLEN
                            self.out_buf[ch][wt_idx] += out_n
                    
                    self.out_hist[ch].fill(0)
                    self.hist_ptrs[ch] = 0
                
                out_wt_ptr += self.HOPSIZE
                if out_wt_ptr >= self.BUFLEN:
                    out_wt_ptr -= self.BUFLEN
        
        # Update pointers
        self.in_rd_ptrs[ch] = in_rd_ptr
        self.in_wt_ptrs[ch] = in_wt_ptr
        self.smp_cnts[ch] = smp_cnt
        self.out_wt_ptrs[ch] = out_wt_ptr
        self.out_rd_ptrs[ch] = out_rd_ptr
        self.ex_ptrs[ch] = ex_ptr
        self.hist_ptrs[ch] = hist_ptr
    
    def set_exlen(self, val: int):
        self.EXLEN = val
    
    def get_exlen(self) -> int:
        return self.EXLEN
    
    def get_max_exlen(self) -> int:
        return self.MAX_EXLEN

File: plugin/libs/lpc.cpp
#include "lpc.h"
#include <fstream>

LPC::LPC(int numChannels) {
    double maxFrameDurS = MAX_FRAME_DUR/1000.0;
    ORDER = MAX_ORDER;
    FRAMELEN = (int)(SAMPLERATE*maxFrameDurS);
    prevFrameLen = FRAMELEN;
    HOPSIZE = FRAMELEN/2;
    totalNumChannels = numChannels;
    inWtPtrs.resize(numChannels);
    inRdPtrs.resize(numChannels);
    smpCnts.resize(numChannels);
    outWtPtrs.resize(numChannels);
    outRdPtrs.resize(numChannels);
    exPtrs.resize(numChannels);
    histPtrs.resize(numChannels);
    exCntPtrs.resize(numChannels);
    for (int ch = 0; ch < numChannels; ch++) {
        inWtPtrs[ch] = 0;
        smpCnts[ch] = 0;
        outWtPtrs[ch] = HOPSIZE;
        outRdPtrs[ch] = 0;
        exPtrs[ch] = 0;
        histPtrs[ch] = 0;
        exCntPtrs[ch] = 0;
        inRdPtrs[ch] = 0;
    }
    phi.resize(ORDER+1);
    alphas.resize(ORDER+1);
    reflectionCoeffs.resize(ORDER);
    orderedInBuf.resize(FRAMELEN);
    orderedScBuf.resize(FRAMELEN);
    window.resize(FRAMELEN);
    inBuf.resize(numChannels);
    outBuf.resize(numChannels);
    out_hist.resize(numChannels);
    scBuf.resize(numChannels);
    for (int ch = 0; ch < numChannels; ch++) {
        inBuf[ch].resize(BUFLEN);
        scBuf[ch].resize(BUFLEN);
        outBuf[ch].resize(BUFLEN);
        out_hist[ch].resize(ORDER);
        for (int i = 0; i < BUFLEN; i++) {
            inBuf[ch][i] = 0.0;
            outBuf[ch][i] = 0.0;
        }
        for (int i = 0; i < ORDER; i++) {
            out_hist[ch][i] = 0.0;
        }
    }
    for (int i = 0; i < FRAMELEN; i++) {
        window[i] = 0.5*(1.0-cos(2.0*M_PI*i/(double)(FRAMELEN-1)));
    }
    for (int i = 0; i < orderedInBuf.size(); i++) {
        orderedInBuf[i] = 0.0;
        orderedScBuf[i] = 0.0;
    }
    for (int i = 0; i < ORDER; i++) {
        phi[i] = 0.0;
    }
    phi[ORDER] = 0.0;
    reset_a();
    DBG("DEBUGGING MODE");
}

//http://www.emptyloop.com/technotes/A%20tutorial%20on%20linear%20prediction%20and%20Levinson-Durbin.pdf
double LPC::levinson_durbin() {
    reset_a();
    double E = phi[0];
    for (int k = 0; k < ORDER; k++) {
        double lbda = 0.0;
        for (int j = 0; j <= k; j++) {
            lbda += alphas[j] * phi[k + 1 - j];
        }
        lbda = -lbda / E;
        reflectionCoeffs[k] = lbda;  // Store reflection coefficient
        // Clamp reflection coefficient for stability
        if (reflectionCoeffs[k] >= 1.0) reflectionCoeffs[k] = 0.999;
        if (reflectionCoeffs[k] <= -1.0) reflectionCoeffs[k] = -0.999;
        int half = (k + 1) / 2;
        for (int n = 0; n <= half; n++) {
            double tmp = alphas[k + 1 - n] + lbda * alphas[n];
            alphas[n] += lbda * alphas[k + 1 - n];
            alphas[k + 1 - n] = tmp;
        }
        E *= (1.0 - lbda * lbda);
    }
    return E;
}

double LPC::autocorrelate(const vector<double>& x, int frameSize, int lag) {
    double res = 0.0;
    for (int n = 0; n < frameSize-lag; n++) {
        res += x[n]*x[n+lag];
    }
    return res;
}

void LPC::prepareToPlay() {
    inWtPtrs.resize(totalNumChannels);
    inRdPtrs.resize(totalNumChannels);
    smpCnts.resize(totalNumChannels);
    outWtPtrs.resize(totalNumChannels);
    outRdPtrs.resize(totalNumChannels);
    exPtrs.resize(totalNumChannels);
    histPtrs.resize(totalNumChannels);
    exCntPtrs.resize(totalNumChannels);
    HOPSIZE = FRAMELEN/2;
    prevFrameLen = FRAMELEN;
    for (int i = 0; i < FRAMELEN; i++) {
        window[i] = 0.5*(1.0-cos(2.0*M_PI*i/(double)(FRAMELEN-1)));
    }
    for (int ch = 0; ch < totalNumChannels; ch++) {
        inWtPtrs[ch] = 0;
        smpCnts[ch] = 0;
        outWtPtrs[ch] = HOPSIZE;
        outRdPtrs[ch] = 0;
        exPtrs[ch] = 0;
        histPtrs[ch] = 0;
        exCntPtrs[ch] = 0;
        inRdPtrs[ch] = 0;
    }
    inBuf.resize(totalNumChannels);
    outBuf.resize(totalNumChannels);
    out_hist.resize(totalNumChannels);
    for (int ch = 0; ch < totalNumChannels; ch++) {
        inBuf[ch].resize(BUFLEN);
        outBuf[ch].resize(BUFLEN);
        out_hist[ch].resize(MAX_ORDER);
    }
}

bool LPC::applyLPC(const float *input, float *output, int numSamples, float lpcMix, float exPercentage, int ch, float exStartPos, const float *sidechain, float previousGain, float currentGain) {
    bool audioWarning = false;
    if (noise == nullptr) {
        return audioWarning;
    }
    inWtPtr = inWtPtrs[ch];
    inRdPtr = inRdPtrs[ch];
    smpCnt = smpCnts[ch];
    outRdPtr = outRdPtrs[ch];
    outWtPtr = (outRdPtr+HOPSIZE)%BUFLEN;
    int exStart = static_cast<int>(exStartPos*EXLEN);
    if (exTypeChanged) {
        exPtrs[ch] = exStart;
        exCntPtrs[ch] = 0;
        histPtrs[ch] = 0;
    }
    if (orderChanged) {
        for (int i = 0; i < out_hist[ch].size(); i++) {
            out_hist[ch][i] = 0;
        }
        histPtrs[ch] = 0;
    }
    exPtr = exPtrs[ch];
    exCntPtr = exCntPtrs[ch];
    histPtr = histPtrs[ch];
    double slope = (currentGain-previousGain)/numSamples;
    for (int s = 0; s < numSamples; s++) {
        inBuf[ch][inWtPtr] = (double)input[s];
        if (sidechain != nullptr) {
            scBuf[ch][inWtPtr] = sidechain[s];
        }
        inWtPtr++;
        if (inWtPtr >= BUFLEN) {
            inWtPtr = 0;
        }
        double out = outBuf[ch][outRdPtr];
        double in = inBuf[ch][(inRdPtr+BUFLEN-FRAMELEN)%BUFLEN];
        inRdPtr++;
        if (inRdPtr >= BUFLEN) {
            inRdPtr = 0;
        }
        double gainFactor = previousGain+slope*(double)s;
        float final_out = lpcMix*gainFactor*out+(1-lpcMix)*in;
        if (isnan(final_out)) {
            audioWarning = true;
            final_out = 0.f;
        }
        else if (fabsf(final_out) > 1.f) {
            audioWarning = true;
            final_out /= (2.f*fabsf(final_out));
        }
        output[s] = final_out;
        outBuf[ch][outRdPtr] = 0;
        outRdPtr++;
        if (outRdPtr >= BUFLEN) {
            outRdPtr = 0;
        }
        smpCnt++;
        if (smpCnt >= HOPSIZE) {
            smpCnt = 0;
            for (int i = 0; i < FRAMELEN; i++) {
                int inBufIdx = (inWtPtr+i-FRAMELEN+BUFLEN)%BUFLEN;
                orderedInBuf[i] = window[i]*inBuf[ch][inBufIdx];
            }
            if (sidechain != nullptr) {
                for (int i = 0; i < FRAMELEN; i++) {
                    int inBufIdx = (inWtPtr+i-FRAMELEN+BUFLEN)%BUFLEN;
                    orderedScBuf[i] = scBuf[ch][inBufIdx];
                }
            }
            for (int lag = 0; lag < ORDER+1; lag++) {
                phi[lag] = autocorrelate(orderedInBuf, FRAMELEN, lag);
            }
            if (phi[0] != 0) {
                double G = sqrt(levinson_durbin());
                if (sidechain == nullptr) {
                    for (int n = 0; n < FRAMELEN; n++) {
                        double ex = (*noise)[exPtr];
                        exCntPtr++;
                        exPtr++;
                        if (exCntPtr >= static_cast<int>(exPercentage*EXLEN)) {
                            exCntPtr = 0;
                            exPtr = exStart;
                        }
                        if (exPtr >= EXLEN) {
                            exPtr = 0;
                        }
                        // Lattice filter synthesis
                        double f = G * ex;
                        for (int i = ORDER - 1; i >= 0; --i) {
                            const double ki = -reflectionCoeffs[i];
                            const double bPrev = out_hist[ch][i];      // ẽ^(i-1)[n-1] - backward delay state

                            // Equation 11.100b: e^(i-1)[n] = e^(i)[n] + k_i * ẽ^(i-1)[n-1]
                            const double fPrev = f + ki * bPrev;
                            // Equation 11.100c: ẽ^(i)[n] = ẽ^(i-1)[n-1] - k_i * e^(i-1)[n]
                            const double bNew = bPrev - ki * fPrev;

                            out_hist[ch][i] = bNew;
                            f = fPrev;
                        }
                        double out_n = f;
                        unsigned long wtIdx = (outWtPtr+n)%BUFLEN;
                        double preOlaVal = outBuf[ch][wtIdx];
                        // Change of frame length can cause OLA to add with unwanted audio
                        // in a correct scenario, OLA with 50% overlap will always be adding with 0s in
                        // the last HOPSIZE-many samples, so just set the last HOPSIZE-many outputs to
                        // be equal to out_n
                        if (n < HOPSIZE) {
                            outBuf[ch][wtIdx] += out_n;
                        }
                        else {
                            outBuf[ch][wtIdx] = out_n;
                        }
                    }
                }
                else {
                    for (int n = 0; n < FRAMELEN; n++) {
                        double ex = orderedScBuf[n];
                        // Lattice filter synthesis
                        double f = G * ex;
                        for (int i = ORDER - 1; i >= 0; --i) {
                            const double ki = -reflectionCoeffs[i];
                            const double bPrev = out_hist[ch][i];      // ẽ^(i-1)[n-1] - backward delay state

                            // Equation 11.100b: e^(i-1)[n] = e^(i)[n] + k_i * ẽ^(i-1)[n-1]
                            const double fPrev = f + ki * bPrev;
                            // Equation 11.100c: ẽ^(i)[n] = ẽ^(i-1)[n-1] - k_i * e^(i-1)[n]
                            const double bNew = bPrev - ki * fPrev;

                            out_hist[ch][i] = bNew;
                            f = fPrev;
                        }
                        double out_n = f;
                        unsigned long wtIdx = (outWtPtr+n)%BUFLEN;
                        double preOlaVal = outBuf[ch][wtIdx];
                        if (n < HOPSIZE) {
                            outBuf[ch][wtIdx] += out_n;
                        }
                        else {
                            outBuf[ch][wtIdx] = out_n;
                        }
                    }
                }
                for (int i = 0; i < out_hist[ch].size(); i++) {
                    out_hist[ch][i] = 0;
                }
                histPtrs[ch] = 0;
            }
            outWtPtr += HOPSIZE;
            if (outWtPtr >= BUFLEN) {
                outWtPtr -= BUFLEN;
            }
        }
    }
    inRdPtrs[ch] = inRdPtr;
    inWtPtrs[ch] = inWtPtr;
    smpCnts[ch] = smpCnt;
    outWtPtrs[ch] = outWtPtr;
    outRdPtrs[ch] = outRdPtr;
    exPtrs[ch] = exPtr;
    exCntPtrs[ch] = exCntPtr;
    histPtrs[ch] = histPtr;
    return audioWarning;
}

void LPC::reset_a() {
    alphas[0] = 1.0;
    for (int i = 1; i < ORDER+1; i++) {
        alphas[i] = 0.0;
    }
}
